File: Scala-002/2024/2024-010/2024-010.scala
import scala.io.Source
import scala.collection.mutable
import java.time.Duration
import java.time.Instant

object AdventOfCode2024_010 {

  val INPUT_FILE = "\\\\vmware-host\\Shared Folders\\C\\advent-of-code-002\\input-files\\2024\\2024-010\\input.txt"

  def readMap(filename: String): Array[Array[Int]] = {
    Source.fromFile(filename)
      .getLines()
      .filter(_.nonEmpty)
      .map(_.trim.map(_.asDigit).toArray)
      .toArray
  }

  def neighbors(r: Int, c: Int, rows: Int, cols: Int): Seq[(Int, Int)] = {
    Seq((-1, 0), (1, 0), (0, -1), (0, 1)).flatMap { case (dr, dc) =>
      val nr = r + dr
      val nc = c + dc
      if (nr >= 0 && nr < rows && nc >= 0 && nc < cols) Some((nr, nc)) else None
    }
  }

  def findTrailheadScores(grid: Array[Array[Int]]): Int = {
    val rows = grid.length
    val cols = if (rows > 0) grid(0).length else 0

    val trailheads = for {
      r <- 0 until rows
      c <- 0 until cols
      if grid(r)(c) == 0
    } yield (r, c)

    var totalScore = 0

    for ((startR, startC) <- trailheads) {
      val visited = mutable.Set[(Int, Int)]()
      val queue = mutable.Queue((startR, startC))
      val reachableNines = mutable.Set[(Int, Int)]()

      visited.add((startR, startC))

      while (queue.nonEmpty) {
        val (r, c) = queue.dequeue()
        val currentHeight = grid(r)(c)

        if (currentHeight == 9) {
          reachableNines.add((r, c))
        } else {
          val nextHeight = currentHeight + 1
          for ((nr, nc) <- neighbors(r, c, rows, cols)) {
            if (!visited.contains((nr, nc)) && grid(nr)(nc) == nextHeight) {
              visited.add((nr, nc))
              queue.enqueue((nr, nc))
            }
          }
        }
      }

      totalScore += reachableNines.size
    }

    totalScore
  }

  def countPaths(r: Int, c: Int, grid: Array[Array[Int]], dp: Array[Array[Option[Int]]], rows: Int, cols: Int): Int = {
    dp(r)(c) match {
      case Some(value) => value
      case None =>
        val currentHeight = grid(r)(c)

        if (currentHeight == 9) {
          dp(r)(c) = Some(1)
          1
        } else {
          val nextHeight = currentHeight + 1
          val totalPaths = neighbors(r, c, rows, cols)
            .filter { case (nr, nc) => grid(nr)(nc) == nextHeight }
            .map { case (nr, nc) => countPaths(nr, nc, grid, dp, rows, cols) }
            .sum

          dp(r)(c) = Some(totalPaths)
          totalPaths
        }
    }
  }

  def calculateTotalRating(grid: Array[Array[Int]]): Int = {
    val rows = grid.length
    val cols = if (rows > 0) grid(0).length else 0

    val trailheads = for {
      r <- 0 until rows
      c <- 0 until cols
      if grid(r)(c) == 0
    } yield (r, c)

    val dp = Array.fill(rows, cols)(Option.empty[Int])

    trailheads.map { case (tr, tc) =>
      countPaths(tr, tc, grid, dp, rows, cols)
    }.sum
  }

  def main(args: Array[String]): Unit = {
    val startTimePart1 = Instant.now()
    val grid = readMap(INPUT_FILE)
    val totalScore = findTrailheadScores(grid)
    val endTimePart1 = Instant.now()

    println(s"Part 1 Result: $totalScore")
    println(f"Time taken for Part 1: ${Duration.between(startTimePart1, endTimePart1).toMillis / 1000.0}%.9f s")

    val startTimePart2 = Instant.now()
    val totalRating = calculateTotalRating(grid)
    val endTimePart2 = Instant.now()

    println(s"Part 2 Result: $totalRating")
    println(f"Time taken for Part 2: ${Duration.between(startTimePart2, endTimePart2).toMillis / 1000.0}%.9f s")
  }
}

File: Node-js/2024/2024-010/2024-010.js
const fs = require('fs');
const path = require('path');

const INPUT_FILE = "\\\\vmware-host\\Shared Folders\\C\\advent-of-code-002\\input-files\\2024\\2024-010\\input.txt";

function readMap(filename) {
    const grid = [];
    const lines = fs.readFileSync(filename, 'utf-8').split('\n');
    for (const line of lines) {
        if (line.trim()) {
            grid.push(line.trim().split('').map(Number));
        }
    }
    return grid;
}

function neighbors(r, c, rows, cols) {
    const directions = [
        [-1, 0], [1, 0], [0, -1], [0, 1]
    ];
    const result = [];
    for (const [dr, dc] of directions) {
        const nr = r + dr;
        const nc = c + dc;
        if (nr >= 0 && nr < rows && nc >= 0 && nc < cols) {
            result.push([nr, nc]);
        }
    }
    return result;
}

function findTrailheadScores(grid) {
    const rows = grid.length;
    const cols = rows > 0 ? grid[0].length : 0;

    const trailheads = [];
    for (let r = 0; r < rows; r++) {
        for (let c = 0; c < cols; c++) {
            if (grid[r][c] === 0) {
                trailheads.push([r, c]);
            }
        }
    }

    let totalScore = 0;

    for (const [startR, startC] of trailheads) {
        const visited = new Set();
        const queue = [[startR, startC]];
        visited.add(`${startR},${startC}`);

        const reachableNines = new Set();

        while (queue.length > 0) {
            const [r, c] = queue.shift();
            const currentHeight = grid[r][c];

            if (currentHeight === 9) {
                reachableNines.add(`${r},${c}`);
            } else {
                const nextHeight = currentHeight + 1;
                for (const [nr, nc] of neighbors(r, c, rows, cols)) {
                    if (!visited.has(`${nr},${nc}`) && grid[nr][nc] === nextHeight) {
                        visited.add(`${nr},${nc}`);
                        queue.push([nr, nc]);
                    }
                }
            }
        }

        totalScore += reachableNines.size;
    }

    return totalScore;
}

function countPaths(r, c, grid, dp, rows, cols) {
    if (dp[r][c] !== -1) {
        return dp[r][c];
    }

    const currentHeight = grid[r][c];

    if (currentHeight === 9) {
        dp[r][c] = 1;
        return 1;
    }

    let totalPaths = 0;
    const nextHeight = currentHeight + 1;
    for (const [nr, nc] of neighbors(r, c, rows, cols)) {
        if (grid[nr][nc] === nextHeight) {
            totalPaths += countPaths(nr, nc, grid, dp, rows, cols);
        }
    }

    dp[r][c] = totalPaths;
    return totalPaths;
}

function calculateTotalRating(grid) {
    const rows = grid.length;
    const cols = rows > 0 ? grid[0].length : 0;

    const trailheads = [];
    for (let r = 0; r < rows; r++) {
        for (let c = 0; c < cols; c++) {
            if (grid[r][c] === 0) {
                trailheads.push([r, c]);
            }
        }
    }

    const dp = Array.from({ length: rows }, () => Array(cols).fill(-1));

    let totalRating = 0;
    for (const [tr, tc] of trailheads) {
        totalRating += countPaths(tr, tc, grid, dp, rows, cols);
    }

    return totalRating;
}

function main() {
    const startTimePart1 = process.hrtime.bigint();
    const grid = readMap(INPUT_FILE);
    const totalScore = findTrailheadScores(grid);
    const endTimePart1 = process.hrtime.bigint();

    console.log(`Part 1 Result: ${totalScore}`);
    console.log(`Time taken for Part 1: ${(Number(endTimePart1 - startTimePart1) / 1e9).toFixed(9)} s`);

    const startTimePart2 = process.hrtime.bigint();
    const totalRating = calculateTotalRating(grid);
    const endTimePart2 = process.hrtime.bigint();

    console.log(`Part 2 Result: ${totalRating}`);
    console.log(`Time taken for Part 2: ${(Number(endTimePart2 - startTimePart2) / 1e9).toFixed(9)} s`);
}

main();
